flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceWithWatermarkExtractor.java

```diff
@@ -0,0 +1,351 @@
+
+  @Rule
+  public final HadoopTableResource sourceTableResource =
+      new HadoopTableResource(
+          TEMPORARY_FOLDER, TestFixtures.DATABASE, TestFixtures.TABLE, TestFixtures.TS_SCHEMA);
+
+  @Test
+  public void testWindowing() throws Exception {
+    GenericAppenderHelper dataAppender = appender();
+    List<Record> expectedRecords = Lists.newArrayList();
+
+    // Generate records with the following pattern:
+    // - File 1 - Later records (Watermark 6000000)
+    //    - Split 1 - 2 records (100, "file_1-recordTs_100"), (103, "file_1-recordTs_103")
+    // - File 2 - First records (Watermark 0)
+    //    - Split 1 - 100 records (0, "file_2-recordTs_0"), (1, "file_2-recordTs_1"),...
+    //    - Split 2 - 100 records (0, "file_2-recordTs_0"), (1, "file_2-recordTs_1"),...
+    // - File 3 - Parallel write for the first records (Watermark 60000)
+    //    - Split 1 - 2 records (1, "file_3-recordTs_1"), (3, "file_3-recordTs_3")
+    List<Record> batch = ImmutableList.of(generateRecord(100, "100"), generateRecord(103, "103"));
+    expectedRecords.addAll(batch);
+    dataAppender.appendToTable(batch);
+
+    batch = Lists.newArrayListWithCapacity(100);
+    for (int i = 0; i < RECORD_NUM_FOR_2_SPLITS; ++i) {
+      batch.add(generateRecord(i % 5, "file_2-recordTs_" + i));
+    }
+    expectedRecords.addAll(batch);
+    dataAppender.appendToTable(batch);
+
+    batch =
+        ImmutableList.of(
+            generateRecord(1, "file_3-recordTs_1"), generateRecord(3, "file_3-recordTs_3"));
+    expectedRecords.addAll(batch);
+    dataAppender.appendToTable(batch);
+
+    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    env.setParallelism(2);
+
+    DataStream<RowData> stream =
+        env.fromSource(
+            sourceBuilder()
+                .streaming(true)
+                .monitorInterval(Duration.ofMillis(10))
+                .streamingStartingStrategy(StreamingStartingStrategy.TABLE_SCAN_THEN_INCREMENTAL)
+                .build(),
+            WatermarkStrategy.<RowData>noWatermarks()
+                .withTimestampAssigner(new RowDataTimestampAssigner()),
+            SOURCE_NAME,
+            TypeInformation.of(RowData.class));
+    DataStream<RowData> windowed =
+        stream
+            .windowAll(TumblingEventTimeWindows.of(Time.minutes(5)))
+            .apply(
+                new AllWindowFunction<RowData, RowData, TimeWindow>() {
+                  @Override
+                  public void apply(
+                      TimeWindow window, Iterable<RowData> values, Collector<RowData> out) {
+                    // Just print all the data to confirm everything has arrived
+                    values.forEach(out::collect);
+                  }
+                });
+
+    try (CloseableIterator<RowData> resultIterator = windowed.collectAsync()) {
+      env.executeAsync("Iceberg Source Windowing Test");
+
+      // Write data so the windows containing test data are closed
```

Comment:
I am not sure how the window and this last records validate watermark alignment. if watermark alignment doesn't work, would it affect the result and assertion.

Reply:
If the watermarks are not emitted, then the windows are not closed, and there will be no result records